// @shapediver/viewer.rendering-engine-threejs.standard/dist/RenderingEngine.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _RenderingEngine_animations;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderingEngine = void 0;
const THREE = __importStar(require("three"));
const tsyringe_1 = require("tsyringe");
const viewer_rendering_engine_camera_engine_1 = require("@shapediver/viewer.rendering-engine.camera-engine");
const viewer_rendering_engine_canvas_engine_1 = require("@shapediver/viewer.rendering-engine.canvas-engine");
const viewer_shared_node_tree_1 = require("@shapediver/viewer.shared.node-tree");
const viewer_rendering_engine_light_engine_1 = require("@shapediver/viewer.rendering-engine.light-engine");
const viewer_rendering_engine_rendering_engine_1 = require("@shapediver/viewer.rendering-engine.rendering-engine");
const viewer_shared_services_1 = require("@shapediver/viewer.shared.services");
const viewer_shared_types_1 = require("@shapediver/viewer.shared.types");
const SceneTreeManager_1 = require("./managers/SceneTreeManager");
const RenderingManager_1 = require("./managers/RenderingManager");
const MaterialLoader_1 = require("./loaders/MaterialLoader");
const EnvironmentMapLoader_1 = require("./loaders/EnvironmentMapLoader");
const GeometryLoader_1 = require("./loaders/GeometryLoader");
const LightLoader_1 = require("./loaders/LightLoader");
const HTMLElementAnchorLoader_1 = require("./loaders/HTMLElementAnchorLoader");
const BeautyRenderingManager_1 = require("./managers/BeautyRenderingManager");
const EnvironmentGeometryManager_1 = require("./managers/EnvironmentGeometryManager");
const SceneTracingManager_1 = require("./managers/SceneTracingManager");
const CameraManager_1 = require("./managers/CameraManager");
const AnimationManager_1 = require("./managers/AnimationManager");
class RenderingEngine {
    constructor(properties) {
        this._converter = tsyringe_1.container.resolve(viewer_shared_services_1.Converter);
        this._logger = tsyringe_1.container.resolve(viewer_shared_services_1.Logger);
        this._canvasEngine = tsyringe_1.container.resolve(viewer_rendering_engine_canvas_engine_1.CanvasEngine);
        this._eventEngine = tsyringe_1.container.resolve(viewer_shared_services_1.EventEngine);
        this._settingsEngine = tsyringe_1.container.resolve(viewer_shared_services_1.SettingsEngine);
        this._stateEngine = tsyringe_1.container.resolve(viewer_shared_services_1.StateEngine);
        this._tree = tsyringe_1.container.resolve(viewer_shared_node_tree_1.Tree);
        this._ambientOcclusion = true;
        this._ambientOcclusionIntensity = 0.1;
        this._automaticResizing = true;
        this._beautyRenderBlendingDuration = 1500;
        this._beautyRenderDelay = 50;
        this._blur = false;
        this._blurSceneWhenBusy = true;
        this._busy = false;
        this._clearAlpha = 1.0;
        this._clearColor = '#ffffff';
        this._environmentMap = 'none';
        this._environmentMapAsBackground = false;
        this._environmentMapResolution = '1024';
        this._gridVisibility = true;
        this._groundPlaneVisibility = true;
        this._lightScene = 'standard';
        this._pointSize = 1.0;
        this._renderingSettings = {
            physicallyCorrectLights: false,
            textureEncoding: THREE.LinearEncoding,
            outputEncoding: THREE.LinearEncoding
        };
        this._shadows = true;
        this._show = false;
        this._showStatistics = false;
        this._closed = false;
        _RenderingEngine_animations.set(this, []);
        THREE.Object3D.DefaultUp = new THREE.Vector3(0, 0, 1);
        this._id = properties.id;
        this._visibility = properties.visibility;
        this._logo = properties.logo;
        this._canvas = this._canvasEngine.getCanvas(this._canvasEngine.createCanvasObject(properties.canvas));
        this._domEventEngine = new viewer_shared_services_1.DomEventEngine(this._id, this.canvas.canvasElement);
        this._cameraEngine = new viewer_rendering_engine_camera_engine_1.CameraEngine(this._id, this.canvas, this._domEventEngine);
        this._lightEngine = new viewer_rendering_engine_light_engine_1.LightEngine(this._id);
        this._animationManager = new AnimationManager_1.AnimationManager(this);
        this._beautyRenderingManager = new BeautyRenderingManager_1.BeautyRenderingManager(this);
        this._cameraManager = new CameraManager_1.CameraManager(this);
        this._environmentGeometryManager = new EnvironmentGeometryManager_1.EnvironmentGeometryManager(this);
        this._sceneTracingManager = new SceneTracingManager_1.SceneTracingManager(this);
        this._sceneTreeManager = new SceneTreeManager_1.SceneTreeManager(this);
        this._renderingManager = new RenderingManager_1.RenderingManager(this);
        this._environmentMapLoader = new EnvironmentMapLoader_1.EnvironmentMapLoader(this);
        this._materialLoader = new MaterialLoader_1.MaterialLoader(this);
        this._geometryLoader = new GeometryLoader_1.GeometryLoader(this);
        this._htmlElementAnchorLoader = new HTMLElementAnchorLoader_1.HTMLElementAnchorLoader(this);
        this._lightLoader = new LightLoader_1.LightLoader(this);
        this._renderer = this.renderingManager.createRenderer(this._canvas.canvasElement);
        this._logoDivElement = this.renderingManager.addLogo(this._canvas.canvasElement, this._logo);
        this._beautyRenderingManager.init();
        this._cameraManager.init();
        this._environmentGeometryManager.init();
        this._sceneTracingManager.init();
        this._sceneTreeManager.init();
        this._renderingManager.init();
        this._environmentMapLoader.init();
        this._materialLoader.init();
        this._geometryLoader.init();
        this._htmlElementAnchorLoader.init();
        this._lightLoader.init();
        this._renderingManager.start();
        this._stateEngine.createCustomState(this.id + '_settings_loaded');
        if (this._visibility === viewer_rendering_engine_rendering_engine_1.VISIBILITYMODE.INSTANT)
            this.show = true;
        if (this._visibility === viewer_rendering_engine_rendering_engine_1.VISIBILITYMODE.SESSION) {
            this._stateEngine.primarySessionLoaded.then(() => {
                if (this._closed)
                    return;
                this._stateEngine.getCustomState(this.id + '_settings_loaded').then(() => {
                    if (this._closed)
                        return;
                    this._environmentGeometryManager.changeSceneExtents(this._sceneTreeManager.boundingBox);
                    this.show = true;
                });
            });
        }
        if (this._stateEngine.primarySettingsRegistered.resolved) {
            if (this._closed)
                return;
            this.applySettings();
        }
        else {
            this._stateEngine.primarySettingsRegistered.then(() => {
                if (this._closed)
                    return;
                this.applySettings();
            });
        }
        this._eventEngine.addListener(viewer_shared_services_1.EVENTTYPE.SETTINGS.SETTINGS_REGISTERED_EXTERNAL, (e) => {
            if (this._closed)
                return;
            this.applySettings();
        });
    }
    get ambientOcclusion() {
        return this._ambientOcclusion;
    }
    set ambientOcclusion(value) {
        this._ambientOcclusion = value;
    }
    get ambientOcclusionIntensity() {
        return this._ambientOcclusionIntensity;
    }
    set ambientOcclusionIntensity(value) {
        this._ambientOcclusionIntensity = value;
    }
    get animationManager() {
        return this._animationManager;
    }
    get animations() {
        return __classPrivateFieldGet(this, _RenderingEngine_animations, "f");
    }
    get automaticResizing() {
        return this._automaticResizing;
    }
    set automaticResizing(value) {
        this._automaticResizing = value;
    }
    get beautyRenderBlendingDuration() {
        return this._beautyRenderBlendingDuration;
    }
    set beautyRenderBlendingDuration(value) {
        this._beautyRenderBlendingDuration = value;
    }
    get beautyRenderDelay() {
        return this._beautyRenderDelay;
    }
    set beautyRenderDelay(value) {
        this._beautyRenderDelay = value;
    }
    get beautyRenderingManager() {
        return this._beautyRenderingManager;
    }
    get blur() {
        return this._blur;
    }
    set blur(value) {
        this._blur = value;
    }
    get blurSceneWhenBusy() {
        return this._blurSceneWhenBusy;
    }
    set blurSceneWhenBusy(value) {
        this._blurSceneWhenBusy = value;
    }
    get busy() {
        return this._busy;
    }
    set busy(value) {
        this._busy = value;
    }
    get cameraEngine() {
        return this._cameraEngine;
    }
    get cameraManager() {
        return this._cameraManager;
    }
    get canvas() {
        return this._canvas;
    }
    get canvasEngine() {
        return this._canvasEngine;
    }
    get clearAlpha() {
        return this._clearAlpha;
    }
    set clearAlpha(value) {
        this._clearAlpha = value;
    }
    get clearColor() {
        return this._clearColor;
    }
    set clearColor(value) {
        this._clearColor = value;
    }
    get closed() {
        return this._closed;
    }
    get domEventEngine() {
        return this._domEventEngine;
    }
    get environmentMap() {
        return this._environmentMap;
    }
    set environmentMap(value) {
        this._environmentMap = value;
        this._environmentMapLoader.load(this.environmentMap);
    }
    get environmentMapAsBackground() {
        return this._environmentMapAsBackground;
    }
    set environmentMapAsBackground(value) {
        this._environmentMapAsBackground = value;
    }
    get environmentMapLoader() {
        return this._environmentMapLoader;
    }
    get environmentMapResolution() {
        return this._environmentMapResolution;
    }
    set environmentMapResolution(value) {
        this._environmentMapResolution = value;
        this._environmentMapLoader.load(this.environmentMap);
    }
    get eventEngine() {
        return this._eventEngine;
    }
    get geometryLoader() {
        return this._geometryLoader;
    }
    get gridVisibility() {
        return this._gridVisibility;
    }
    set gridVisibility(value) {
        if (this._environmentGeometryManager.grid)
            this._environmentGeometryManager.grid.visible = value;
        this._gridVisibility = value;
    }
    get groundPlaneVisibility() {
        return this._groundPlaneVisibility;
    }
    set groundPlaneVisibility(value) {
        if (this._environmentGeometryManager.groundPlane)
            this._environmentGeometryManager.groundPlane.visible = value;
        this._groundPlaneVisibility = value;
    }
    get htmlElementAnchorLoader() {
        return this._htmlElementAnchorLoader;
    }
    get id() {
        return this._id;
    }
    get lightEngine() {
        return this._lightEngine;
    }
    get lightSceneId() {
        return this._lightEngine.lightScene.id;
    }
    get lightLoader() {
        return this._lightLoader;
    }
    get lightScene() {
        return this._lightScene;
    }
    set lightScene(value) {
        this._lightScene = value;
    }
    get logoDivElement() {
        return this._logoDivElement;
    }
    get materialLoader() {
        return this._materialLoader;
    }
    get minimalRendering() {
        return this.renderingManager.minimalRendering;
    }
    get pointSize() {
        return this._pointSize;
    }
    set pointSize(value) {
        this._pointSize = value;
        this.materialLoader.assignPointSize(value);
    }
    get renderer() {
        return this._renderer;
    }
    get renderingManager() {
        return this._renderingManager;
    }
    get renderingSettings() {
        return this._renderingSettings;
    }
    set renderingSettings(value) {
        this._renderingSettings = value;
        if (value.envMapIntensity !== undefined)
            this._materialLoader.assignEnvironmentMapIntensity(value.envMapIntensity);
        if (value.envMapIntensityGroundPlane !== undefined)
            this._environmentGeometryManager.assignGroundPlaneEnvironmentIntensity(value.envMapIntensityGroundPlane);
        if (value.groundPlaneColor !== undefined)
            this._environmentGeometryManager.assignGroundPlaneColor(value.groundPlaneColor);
        if (value.toneMapping !== undefined)
            this._renderer.toneMapping = value.toneMapping;
        if (value.toneMappingExposure !== undefined)
            this._renderer.toneMappingExposure = value.toneMappingExposure;
        if (value.physicallyCorrectLights !== undefined)
            this._renderer.physicallyCorrectLights = value.physicallyCorrectLights;
        if (value.outputEncoding !== undefined) {
            this._renderer.outputEncoding = value.outputEncoding;
            if (value.outputEncoding === 3000 || value.outputEncoding === 3001) {
                this._beautyRenderingManager.assignOutputEncoding(value.outputEncoding);
            }
            else {
                this._logger.warn(viewer_shared_services_1.LOGGINGTOPIC.VIEWER, 'Output encoding of this type cannot be used in combination with Ambient Occlusion at the moment.');
            }
        }
        if (value.textureEncoding !== undefined)
            this._materialLoader.assignTextureEncoding(value.textureEncoding);
    }
    get scene() {
        return this._sceneTreeManager.scene;
    }
    get sceneTracingManager() {
        return this._sceneTracingManager;
    }
    get sceneTreeManager() {
        return this._sceneTreeManager;
    }
    get settingsEngine() {
        return this._settingsEngine;
    }
    get shadows() {
        return this._shadows;
    }
    set shadows(value) {
        this._shadows = value;
    }
    get show() {
        return this._show;
    }
    set show(value) {
        this._show = value;
    }
    get showStatistics() {
        return this._showStatistics;
    }
    set showStatistics(value) {
        this._showStatistics = value;
    }
    get stateEngine() {
        return this._stateEngine;
    }
    get usingSwiftShader() {
        return this.renderingManager.usingSwiftShader;
    }
    gatherAnimations(node = this._tree.root) {
        let out = [];
        for (let i = 0, len = node.data.length; i < len; i++)
            if (node.data[i] instanceof viewer_shared_types_1.AnimationData)
                out.push(node.data[i]);
        for (let i = 0, len = node.children.length; i < len; i++)
            out = out.concat(this.gatherAnimations(node.children[i]));
        return out;
    }
    close() {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            this._closed = true;
            (_a = this._canvas.canvasElement.parentElement) === null || _a === void 0 ? void 0 : _a.removeChild(this._logoDivElement);
            (_b = this._canvas.canvasElement.parentNode) === null || _b === void 0 ? void 0 : _b.removeChild(this._htmlElementAnchorLoader.parentDiv);
            this._canvas.reset();
            this._domEventEngine.removeAllDomEventListener();
            this._domEventEngine.dispose();
            return true;
        });
    }
    getScreenshot(type, encoderOptions) {
        return this._renderingManager.getScreenshot(type, encoderOptions);
    }
    reset() {
        this._environmentGeometryManager.changeSceneExtents(this._sceneTreeManager.boundingBox);
        if (this._visibility === viewer_rendering_engine_rendering_engine_1.VISIBILITYMODE.SESSION)
            this.show = false;
        this._stateEngine.getCustomState(this.id + '_settings_loaded').reset();
    }
    resize(width, height) {
        this._renderingManager.resize(width, height);
        this._renderingManager.render();
    }
    saveSettings() {
        this.lightEngine.saveSettings();
        this.cameraEngine.saveSettings();
        this._settingsEngine.general.blurWhenBusy = this.blurSceneWhenBusy;
        this._settingsEngine.environmentGeometry.gridVisibility = this.gridVisibility;
        this._settingsEngine.environmentGeometry.groundPlaneVisibility = this.groundPlaneVisibility;
        this._settingsEngine.light.lightSceneId = this.lightScene;
        this._settingsEngine.environment.mapResolution = this.environmentMapResolution;
        this._settingsEngine.environment.map = Array.isArray(this.environmentMap) ? JSON.stringify(this.environmentMap) : this.environmentMap;
        this._settingsEngine.environment.mapAsBackground = this.environmentMapAsBackground;
        this._settingsEngine.rendering.ambientOcclusion = this.ambientOcclusion;
        this._settingsEngine.rendering.ambientOcclusionIntensity = this.ambientOcclusionIntensity;
        this._settingsEngine.rendering.beautyRenderBlendingDuration = this.beautyRenderBlendingDuration;
        this._settingsEngine.rendering.beautyRenderDelay = this.beautyRenderDelay;
        this._settingsEngine.environment.clearAlpha = this.clearAlpha;
        this._settingsEngine.environment.clearColor = this.clearColor;
        this._settingsEngine.general.pointSize = this.pointSize;
        this._settingsEngine.rendering.shadows = this.shadows;
    }
    update() {
        this._sceneTreeManager.updateSceneTree(this._tree.root, this._lightEngine);
        this._renderingManager.updateShadowMap();
        __classPrivateFieldSet(this, _RenderingEngine_animations, this.gatherAnimations(), "f");
        this._renderingManager.render();
    }
    applySettings() {
        const token = this._eventEngine.addListener(viewer_shared_services_1.EVENTTYPE.ENVIRONMENTMAP.ENVIRONMENTMAP_LOADED, (e) => {
            const viewerEvent = e;
            if (viewerEvent.viewerId !== this.id)
                return;
            this._eventEngine.removeListener(token);
            if (!viewerEvent.environmentMapId || (viewerEvent.environmentMapId && !Array.isArray(this._settingsEngine.environment.map) && viewerEvent.environmentMapId !== this._settingsEngine.environment.map.toLowerCase()))
                return;
            this.environmentMapAsBackground = this._settingsEngine.environment.mapAsBackground;
            this.ambientOcclusion = this._settingsEngine.rendering.ambientOcclusion;
            this.ambientOcclusionIntensity = this._settingsEngine.rendering.ambientOcclusionIntensity;
            this.beautyRenderBlendingDuration = this._settingsEngine.rendering.beautyRenderBlendingDuration;
            this.beautyRenderDelay = this._settingsEngine.rendering.beautyRenderDelay;
            this.blurSceneWhenBusy = this._settingsEngine.general.blurWhenBusy;
            this.clearAlpha = this._settingsEngine.environment.clearAlpha;
            this.clearColor = this._converter.toColor(this._settingsEngine.environment.clearColor);
            this.gridVisibility = this._settingsEngine.environmentGeometry.gridVisibility;
            this.groundPlaneVisibility = this._settingsEngine.environmentGeometry.groundPlaneVisibility;
            this.lightScene = this._settingsEngine.light.lightSceneId;
            this.pointSize = this._settingsEngine.general.pointSize;
            this.shadows = this._settingsEngine.rendering.shadows;
            this.lightEngine.applySettings();
            this.cameraEngine.applySettings();
            this._stateEngine.getCustomState(this.id + '_settings_loaded').resolve(true);
            this.update();
        });
        this._environmentMapResolution = this._settingsEngine.environment.mapResolution;
        this.environmentMap = this._settingsEngine.environment.map;
    }
}
exports.RenderingEngine = RenderingEngine;
_RenderingEngine_animations = new WeakMap();
