// @shapediver/api.geometry-api-dto-v1/dist/request/SdRequestDtoModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverRequestModelTrustlevel = exports.ShapeDiverRequestModelFiletype = exports.ShapeDiverRequestModelTemplateDataName = void 0;
var ShapeDiverRequestModelTemplateDataName;
(function (ShapeDiverRequestModelTemplateDataName) {
    ShapeDiverRequestModelTemplateDataName["ACCESSDOMAINS"] = "accessdomains";
    ShapeDiverRequestModelTemplateDataName["AUTH_GROUPS"] = "auth_groups";
    ShapeDiverRequestModelTemplateDataName["BACKENDACCESS"] = "backendaccess";
    ShapeDiverRequestModelTemplateDataName["DENY_SCRIPTS"] = "deny_scripts";
    ShapeDiverRequestModelTemplateDataName["FILENAME"] = "filename";
    ShapeDiverRequestModelTemplateDataName["FTYPE"] = "ftype";
    ShapeDiverRequestModelTemplateDataName["ID2"] = "id2";
    ShapeDiverRequestModelTemplateDataName["INITIAL_WARMUP"] = "initial_warmup";
    ShapeDiverRequestModelTemplateDataName["MAX_COMP_TIME"] = "max_comp_time";
    ShapeDiverRequestModelTemplateDataName["MAX_EXPORT_SIZE"] = "max_export_size";
    ShapeDiverRequestModelTemplateDataName["MAX_IDLE_MINUTES"] = "max_idle_minutes";
    ShapeDiverRequestModelTemplateDataName["MAX_MODEL_SIZE"] = "max_model_size";
    ShapeDiverRequestModelTemplateDataName["MAX_OUTPUT_SIZE"] = "max_output_size";
    ShapeDiverRequestModelTemplateDataName["MAX_WAIT_TIME"] = "max_wait_time";
    ShapeDiverRequestModelTemplateDataName["NAME"] = "name";
    ShapeDiverRequestModelTemplateDataName["NUM_LOADED_MAX"] = "num_loaded_max";
    ShapeDiverRequestModelTemplateDataName["NUM_LOADED_MIN"] = "num_loaded_min";
    ShapeDiverRequestModelTemplateDataName["NUM_PRELOADED_MIN"] = "num_preloaded_min";
    ShapeDiverRequestModelTemplateDataName["PREV_ID"] = "prev_id";
    ShapeDiverRequestModelTemplateDataName["PUB"] = "pub";
    ShapeDiverRequestModelTemplateDataName["REQUIRE_TOKEN"] = "require_token";
    ShapeDiverRequestModelTemplateDataName["SESSION_RATE_LIMIT"] = "session_rate_limit";
    ShapeDiverRequestModelTemplateDataName["TRUST"] = "trust";
    ShapeDiverRequestModelTemplateDataName["WEBHOOK_TOKEN"] = "webhook_token";
    ShapeDiverRequestModelTemplateDataName["WEBHOOK_URL"] = "webhook_url";
})(ShapeDiverRequestModelTemplateDataName = exports.ShapeDiverRequestModelTemplateDataName || (exports.ShapeDiverRequestModelTemplateDataName = {}));
var ShapeDiverRequestModelFiletype;
(function (ShapeDiverRequestModelFiletype) {
    ShapeDiverRequestModelFiletype["UNKNOWN"] = "unknown";
    ShapeDiverRequestModelFiletype["GRASSHOPPER_BINARY"] = "gh";
    ShapeDiverRequestModelFiletype["GRASSHOPPER_XML"] = "ghx";
})(ShapeDiverRequestModelFiletype = exports.ShapeDiverRequestModelFiletype || (exports.ShapeDiverRequestModelFiletype = {}));
var ShapeDiverRequestModelTrustlevel;
(function (ShapeDiverRequestModelTrustlevel) {
    ShapeDiverRequestModelTrustlevel["UNDEFINED"] = "";
    ShapeDiverRequestModelTrustlevel["NONE"] = "none";
    ShapeDiverRequestModelTrustlevel["FULL"] = "full";
})(ShapeDiverRequestModelTrustlevel = exports.ShapeDiverRequestModelTrustlevel || (exports.ShapeDiverRequestModelTrustlevel = {}));
