// @shapediver/api.geometry-api-dto-v1/dist/request/SdRequestDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverRequestGltfUploadQueryConversion = exports.ShapeDiverRequestLogMessageLevel = exports.ShapeDiverRequestTicketType = void 0;
var ShapeDiverRequestTicketType;
(function (ShapeDiverRequestTicketType) {
    ShapeDiverRequestTicketType["BACKEND"] = "backend";
    ShapeDiverRequestTicketType["NONE"] = "";
})(ShapeDiverRequestTicketType = exports.ShapeDiverRequestTicketType || (exports.ShapeDiverRequestTicketType = {}));
var ShapeDiverRequestLogMessageLevel;
(function (ShapeDiverRequestLogMessageLevel) {
    ShapeDiverRequestLogMessageLevel[ShapeDiverRequestLogMessageLevel["INFO"] = 0] = "INFO";
    ShapeDiverRequestLogMessageLevel[ShapeDiverRequestLogMessageLevel["WARNING"] = 1] = "WARNING";
    ShapeDiverRequestLogMessageLevel[ShapeDiverRequestLogMessageLevel["ERROR"] = 2] = "ERROR";
})(ShapeDiverRequestLogMessageLevel = exports.ShapeDiverRequestLogMessageLevel || (exports.ShapeDiverRequestLogMessageLevel = {}));
var ShapeDiverRequestGltfUploadQueryConversion;
(function (ShapeDiverRequestGltfUploadQueryConversion) {
    ShapeDiverRequestGltfUploadQueryConversion["NONE"] = "none";
    ShapeDiverRequestGltfUploadQueryConversion["USDZ"] = "usdz";
})(ShapeDiverRequestGltfUploadQueryConversion = exports.ShapeDiverRequestGltfUploadQueryConversion || (exports.ShapeDiverRequestGltfUploadQueryConversion = {}));
