// @shapediver/viewer.settings/dist/versions/v2/Validator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validate = void 0;
const joi_1 = __importDefault(require("joi"));
const lightSchema = joi_1.default.object({
    id: joi_1.default.string(),
    name: joi_1.default.string().optional(),
    type: joi_1.default.string(),
    order: joi_1.default.number().optional(),
    properties: {
        color: joi_1.default.alternatives([joi_1.default.number(), joi_1.default.string()]).optional(),
        direction: joi_1.default.object({ x: joi_1.default.number(), y: joi_1.default.number(), z: joi_1.default.number() }).optional(),
        position: joi_1.default.object({ x: joi_1.default.number(), y: joi_1.default.number(), z: joi_1.default.number() }).optional(),
        target: joi_1.default.object({ x: joi_1.default.number(), y: joi_1.default.number(), z: joi_1.default.number() }).optional(),
        castShadow: joi_1.default.boolean().optional(),
        skyColor: joi_1.default.alternatives([joi_1.default.number(), joi_1.default.string()]).optional(),
        groundColor: joi_1.default.alternatives([joi_1.default.number(), joi_1.default.string()]).optional(),
        intensity: joi_1.default.number().optional(),
        distance: joi_1.default.number().optional(),
        angle: joi_1.default.number().optional(),
        penumbra: joi_1.default.number().optional(),
        decay: joi_1.default.number().optional(),
        shadowMapResolution: joi_1.default.number().optional(),
        shadowMapBias: joi_1.default.number().optional()
    }
});
const lightScenesSchema = joi_1.default.object().pattern(/^/, [
    joi_1.default.object({
        id: joi_1.default.string(),
        name: joi_1.default.string().optional(),
        lights: joi_1.default.object().pattern(/^/, [lightSchema])
    })
]);
const schema = joi_1.default.object({
    build_date: joi_1.default.string().allow('').optional(),
    build_version: joi_1.default.string().allow('').optional(),
    settings_version: joi_1.default.string(),
    ar: {
        enableCameraSync: joi_1.default.boolean().optional(),
        enableCameraSyncInitial: joi_1.default.boolean().optional(),
        enableLightingEstimation: joi_1.default.boolean().optional(),
        enableTouchControls: joi_1.default.boolean().optional(),
        enableTouchControlRotation: joi_1.default.boolean().optional(),
        enableAutomaticPlacement: joi_1.default.boolean().optional(),
        defaultHitTestType: joi_1.default.string().optional(),
    },
    defaultMaterial: {
        bumpAmplitude: joi_1.default.number().optional(),
        color: joi_1.default.string().optional(),
        metalness: joi_1.default.number().optional(),
        roughness: joi_1.default.number().optional(),
    },
    parameters: {
        controlOrder: joi_1.default.array().items(joi_1.default.string()),
        controlNames: joi_1.default.object().pattern(/.*/, [joi_1.default.string()]),
        parametersHidden: joi_1.default.array().items(joi_1.default.string()),
    },
    viewer: {
        blurSceneWhenBusy: joi_1.default.boolean(),
        ignoreSuperseded: joi_1.default.boolean().optional(),
        loggingLevel: joi_1.default.number().optional(),
        messageLoggingLevel: joi_1.default.number().optional(),
        viewerRuntimeId: joi_1.default.string().allow('').optional(),
        hasRestoredSettings: joi_1.default.boolean().optional(),
        useModelSettings: joi_1.default.boolean().optional(),
        showMessages: joi_1.default.boolean().optional(),
        commitSettings: joi_1.default.boolean(),
        commitParameters: joi_1.default.boolean(),
        scene: {
            show: joi_1.default.boolean().optional(),
            showSceneTransition: joi_1.default.string().optional(),
            duration: joi_1.default.number().optional(),
            fullscreen: joi_1.default.boolean().optional(),
            gridVisibility: joi_1.default.boolean(),
            groundPlaneReflectionThreshold: joi_1.default.number().optional(),
            groundPlaneReflectionVisibility: joi_1.default.boolean().optional(),
            groundPlaneVisibility: joi_1.default.boolean(),
            camera: {
                autoAdjust: joi_1.default.boolean(),
                cameraMovementDuration: joi_1.default.number().min(0),
                cameraTypes: {
                    perspective: {
                        default: { position: { x: joi_1.default.number(), y: joi_1.default.number(), z: joi_1.default.number() }, target: { x: joi_1.default.number(), y: joi_1.default.number(), z: joi_1.default.number() } },
                        fov: joi_1.default.number().min(0).max(180),
                        controls: joi_1.default.number().optional(),
                    },
                    orthographic: {
                        default: { position: { x: joi_1.default.number(), y: joi_1.default.number(), z: joi_1.default.number() }, target: { x: joi_1.default.number(), y: joi_1.default.number(), z: joi_1.default.number() } },
                    },
                    active: joi_1.default.number().min(0).max(6)
                },
                controls: {
                    orbit: {
                        autoRotationSpeed: joi_1.default.number(),
                        damping: joi_1.default.number().min(0),
                        enableAutoRotation: joi_1.default.boolean(),
                        enableKeyPan: joi_1.default.boolean(),
                        enablePan: joi_1.default.boolean(),
                        enableRotation: joi_1.default.boolean(),
                        enableZoom: joi_1.default.boolean(),
                        input: { keys: { up: joi_1.default.number(), down: joi_1.default.number(), left: joi_1.default.number(), right: joi_1.default.number() }, mouse: { rotate: joi_1.default.number(), zoom: joi_1.default.number(), pan: joi_1.default.number() }, touch: { rotate: joi_1.default.number(), zoom: joi_1.default.number(), pan: joi_1.default.number() }, },
                        keyPanSpeed: joi_1.default.number().min(0).max(1),
                        movementSmoothness: joi_1.default.number().min(0).max(1),
                        restrictions: {
                            position: joi_1.default.object({
                                cube: joi_1.default.object({ min: { x: joi_1.default.number().allow(-Infinity), y: joi_1.default.number().allow(-Infinity), z: joi_1.default.number().allow(-Infinity) }, max: { x: joi_1.default.number().allow(Infinity), y: joi_1.default.number().allow(Infinity), z: joi_1.default.number().allow(Infinity) } }).optional(),
                                sphere: joi_1.default.object({ center: { x: joi_1.default.number(), y: joi_1.default.number(), z: joi_1.default.number() }, radius: joi_1.default.number().allow(Infinity) }).optional(),
                            }).optional(),
                            target: joi_1.default.object({
                                cube: joi_1.default.object({ min: { x: joi_1.default.number().allow(-Infinity), y: joi_1.default.number().allow(-Infinity), z: joi_1.default.number().allow(-Infinity) }, max: { x: joi_1.default.number().allow(Infinity), y: joi_1.default.number().allow(Infinity), z: joi_1.default.number().allow(Infinity) } }).optional(),
                                sphere: joi_1.default.object({ center: { x: joi_1.default.number(), y: joi_1.default.number(), z: joi_1.default.number() }, radius: joi_1.default.number().allow(Infinity) }).optional(),
                            }).optional(),
                            rotation: joi_1.default.object({ minPolarAngle: joi_1.default.number(), maxPolarAngle: joi_1.default.number(), minAzimuthAngle: joi_1.default.number().allow(-Infinity), maxAzimuthAngle: joi_1.default.number().allow(Infinity) }).optional(),
                            zoom: joi_1.default.object({ minDistance: joi_1.default.number(), maxDistance: joi_1.default.number().allow(Infinity) }).optional(),
                        },
                        rotationSpeed: joi_1.default.number().min(0).max(1),
                        panSpeed: joi_1.default.number().min(0).max(1),
                        zoomSpeed: joi_1.default.number().min(0).max(1),
                    },
                    fps: {},
                    orthographic: {
                        damping: joi_1.default.number().min(0).max(1),
                        enableKeyPan: joi_1.default.boolean(),
                        enablePan: joi_1.default.boolean(),
                        enableZoom: joi_1.default.boolean(),
                        input: { keys: { up: joi_1.default.number(), down: joi_1.default.number(), left: joi_1.default.number(), right: joi_1.default.number() }, mouse: { rotate: joi_1.default.number(), zoom: joi_1.default.number(), pan: joi_1.default.number() }, touch: { rotate: joi_1.default.number(), zoom: joi_1.default.number(), pan: joi_1.default.number() }, },
                        keyPanSpeed: joi_1.default.number().min(0).max(1),
                        movementSmoothness: joi_1.default.number().min(0).max(1),
                        panSpeed: joi_1.default.number().min(0).max(1),
                        zoomSpeed: joi_1.default.number().min(0).max(1),
                    }
                },
                enableCameraControls: joi_1.default.boolean(),
                revertAtMouseUp: joi_1.default.boolean(),
                revertAtMouseUpDuration: joi_1.default.number().min(0),
                zoomExtentsFactor: joi_1.default.number().positive(),
            },
            lights: {
                helper: joi_1.default.boolean().optional(),
                lightScene: joi_1.default.string().allow(''),
                lightScenes: lightScenesSchema.allow(null)
            },
            material: {
                environmentMap: joi_1.default.alternatives([joi_1.default.string().allow(''), joi_1.default.array().items(joi_1.default.string())]),
                environmentMapAsBackground: joi_1.default.boolean(),
                environmentMapResolution: joi_1.default.string().valid('256', '512', '1024', '2048')
            },
            render: {
                ambientOcclusion: joi_1.default.boolean(),
                beautyRenderDelay: joi_1.default.number().min(0),
                beautyRenderBlendingDuration: joi_1.default.number().min(0),
                clearAlpha: joi_1.default.number().min(0).max(1),
                clearColor: joi_1.default.string(),
                pointSize: joi_1.default.number().min(0),
                shadows: joi_1.default.boolean(),
                sao: {
                    samples: joi_1.default.number().positive().optional(),
                    kernelRadius: joi_1.default.number().positive().optional(),
                    intensity: joi_1.default.number().positive().optional(),
                    standardDev: joi_1.default.number().min(0).optional(),
                }
            },
        },
    },
}).options({ presence: 'required' });
const validate = (s) => {
    const result = joi_1.default.attempt(s, schema);
    s = result.value;
};
exports.validate = validate;
