// @shapediver/viewer.data-engine.geometry-engine/dist/GeometryEngine.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeometryEngine = void 0;
const viewer_shared_node_tree_1 = require("@shapediver/viewer.shared.node-tree");
const tsyringe_1 = require("tsyringe");
const viewer_shared_services_1 = require("@shapediver/viewer.shared.services");
const GLTFLoader_1 = require("./gltfv1/GLTFLoader");
const GLTFLoader_2 = require("./gltfv2/GLTFLoader");
const GLTFConverter_1 = require("./gltfv2/GLTFConverter");
let GeometryEngine = class GeometryEngine {
    constructor() {
        this.BINARY_EXTENSION_HEADER_LENGTH = 20;
        this._httpClient = tsyringe_1.container.resolve(viewer_shared_services_1.HttpClient);
        this._logger = tsyringe_1.container.resolve(viewer_shared_services_1.Logger);
        this._performanceEvaluator = tsyringe_1.container.resolve(viewer_shared_services_1.PerformanceEvaluator);
    }
    loadContent(content) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!content || (content && !content.href)) {
                this._logger.error(viewer_shared_services_1.LOGGINGTOPIC.DATAPROCESSING, new viewer_shared_services_1.SDError('GeometryEngine.loadContent: Invalid content was provided to geometry engine.'), '', false);
                return new viewer_shared_node_tree_1.TreeNode();
            }
            const url = content.href;
            let gltfContent, gltfBinary, gltfBaseUrl, gltfHeader;
            let version = '2.0';
            if (content.format === 'glb' || content.format === 'gltf') {
                this._performanceEvaluator.startSection('gltfProcessing.' + url);
                let axiosResponse;
                try {
                    this._performanceEvaluator.startSection('loadGltf.' + url);
                    axiosResponse = yield this._httpClient.get(url, {
                        responseType: 'arraybuffer'
                    });
                    this._performanceEvaluator.endSection('loadGltf.' + url);
                }
                catch (e) {
                    if (e.response && e.response.status) {
                        this._logger.httpError(viewer_shared_services_1.LOGGINGTOPIC.DATAPROCESSING, e, `GeometryEngine.loadContent: Initial loading of geometry failed.`, e.response.status, false);
                    }
                    else {
                        this._logger.error(viewer_shared_services_1.LOGGINGTOPIC.DATAPROCESSING, e, `GeometryEngine.loadContent: Initial loading of geometry failed.`, false);
                    }
                    return new viewer_shared_node_tree_1.TreeNode();
                }
                const magic = new TextDecoder().decode(new Uint8Array(axiosResponse.data, 0, 4));
                const isBinary = magic === 'glTF' || (axiosResponse.headers['content-type'] &&
                    (axiosResponse.headers['content-type'] === 'model/gltf-binary' ||
                        axiosResponse.headers['content-type'] === 'application/octet-stream' ||
                        axiosResponse.headers['content-type'] === 'model/gltf.binary'));
                if (isBinary) {
                    gltfBinary = axiosResponse.data;
                    const headerDataView = new DataView(gltfBinary, 0, this.BINARY_EXTENSION_HEADER_LENGTH);
                    gltfHeader = {
                        magic: magic,
                        version: headerDataView.getUint32(4, true),
                        length: headerDataView.getUint32(8, true),
                        contentLength: headerDataView.getUint32(12, true),
                        contentFormat: headerDataView.getUint32(16, true)
                    };
                    if (gltfHeader.magic != 'glTF') {
                        this._logger.error(viewer_shared_services_1.LOGGINGTOPIC.DATAPROCESSING, new viewer_shared_services_1.SDError('GeometryEngine.loadContent: Invalid data: glTF magic wrong.'));
                        return new viewer_shared_node_tree_1.TreeNode();
                    }
                    const contentDataView = new DataView(gltfBinary, this.BINARY_EXTENSION_HEADER_LENGTH, gltfHeader.contentLength);
                    const contentDecoded = new TextDecoder().decode(contentDataView);
                    gltfContent = JSON.parse(contentDecoded);
                    if (gltfContent && gltfContent.asset && gltfContent.asset.version) {
                        const assetVersion = (gltfContent.asset.version + '').endsWith('.0') ? gltfContent.asset.version : gltfContent.asset.version + '.0';
                        if (gltfHeader.version + '.0' === assetVersion) {
                            version = gltfHeader.version + '.0';
                        }
                        else {
                            this._logger.error(viewer_shared_services_1.LOGGINGTOPIC.DATAPROCESSING, new viewer_shared_services_1.SDError('GeometryEngine.loadContent: glTF header version (' + gltfHeader.version + ') is not the same as asset version (' + assetVersion + ').'));
                            return new viewer_shared_node_tree_1.TreeNode();
                        }
                    }
                    else {
                        version = gltfHeader.version + '.0';
                    }
                }
                else {
                    gltfContent = JSON.parse(new TextDecoder().decode(axiosResponse.data));
                    if (gltfContent && gltfContent.asset && gltfContent.asset.version) {
                        if (gltfContent.asset.version !== '2.0') {
                            this._logger.error(viewer_shared_services_1.LOGGINGTOPIC.DATAPROCESSING, new viewer_shared_services_1.SDError('GeometryEngine.loadContent: Only gltf v2 is supported in a non-binary format.'));
                            return new viewer_shared_node_tree_1.TreeNode();
                        }
                    }
                    else {
                        this._logger.warn(viewer_shared_services_1.LOGGINGTOPIC.DATAPROCESSING, 'GeometryEngine.loadContent: No version specified in asset, trying to load as v2.');
                        version = '2.0';
                    }
                    const removeLastDirectoryPartOf = (the_url) => {
                        const dir_char = the_url.includes("/") ? "/" : "\\";
                        const the_arr = the_url.split(dir_char);
                        the_arr.pop();
                        return the_arr.join(dir_char);
                    };
                    gltfBaseUrl = removeLastDirectoryPartOf(url);
                    if (!gltfBaseUrl && window && window.location && window.location.href)
                        gltfBaseUrl = removeLastDirectoryPartOf(window.location.href);
                }
            }
            let node;
            if (version === '1.0') {
                node = yield new GLTFLoader_1.GLTFLoader().load(gltfContent, gltfBinary, gltfHeader, gltfBaseUrl);
            }
            else {
                node = yield new GLTFLoader_2.GLTFLoader().load(gltfContent, gltfBinary, gltfHeader, gltfBaseUrl);
            }
            this._performanceEvaluator.endSection('gltfProcessing.' + url);
            return node;
        });
    }
    convertSceneToGLTF(node) {
        return __awaiter(this, void 0, void 0, function* () {
            return new GLTFConverter_1.GLTFConverter().convert(node);
        });
    }
};
GeometryEngine = __decorate([
    tsyringe_1.singleton(),
    __metadata("design:paramtypes", [])
], GeometryEngine);
exports.GeometryEngine = GeometryEngine;
