// @shapediver/viewer.rendering-engine.light-engine/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpotLight = exports.PointLight = exports.HemisphereLight = exports.DirectionalLight = exports.AmbientLight = exports.AbstractLight = exports.LightScene = exports.LIGHTTYPE = exports.LightEngine = void 0;
const AbstractLight_1 = require("./implementation/AbstractLight");
Object.defineProperty(exports, "AbstractLight", { enumerable: true, get: function () { return AbstractLight_1.AbstractLight; } });
const LightEngine_1 = require("./implementation/LightEngine");
Object.defineProperty(exports, "LightEngine", { enumerable: true, get: function () { return LightEngine_1.LightEngine; } });
const LightScene_1 = require("./implementation/LightScene");
Object.defineProperty(exports, "LightScene", { enumerable: true, get: function () { return LightScene_1.LightScene; } });
const AmbientLight_1 = require("./implementation/types/AmbientLight");
Object.defineProperty(exports, "AmbientLight", { enumerable: true, get: function () { return AmbientLight_1.AmbientLight; } });
const DirectionalLight_1 = require("./implementation/types/DirectionalLight");
Object.defineProperty(exports, "DirectionalLight", { enumerable: true, get: function () { return DirectionalLight_1.DirectionalLight; } });
const HemisphereLight_1 = require("./implementation/types/HemisphereLight");
Object.defineProperty(exports, "HemisphereLight", { enumerable: true, get: function () { return HemisphereLight_1.HemisphereLight; } });
const PointLight_1 = require("./implementation/types/PointLight");
Object.defineProperty(exports, "PointLight", { enumerable: true, get: function () { return PointLight_1.PointLight; } });
const SpotLight_1 = require("./implementation/types/SpotLight");
Object.defineProperty(exports, "SpotLight", { enumerable: true, get: function () { return SpotLight_1.SpotLight; } });
const ILight_1 = require("./interface/ILight");
Object.defineProperty(exports, "LIGHTTYPE", { enumerable: true, get: function () { return ILight_1.LIGHTTYPE; } });
