// @shapediver/viewer.shared.services/dist/event-engine/EventTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVENTTYPE = exports.SETTINGS = exports.SESSION = exports.VIEWER = exports.SCENE = exports.RENDERING = exports.ENVIRONMENTMAP = exports.CAMERA = void 0;
var CAMERA;
(function (CAMERA) {
    CAMERA["CAMERA_START"] = "camera.start";
    CAMERA["CAMERA_MOVE"] = "camera.move";
    CAMERA["CAMERA_END"] = "camera.end";
})(CAMERA = exports.CAMERA || (exports.CAMERA = {}));
var ENVIRONMENTMAP;
(function (ENVIRONMENTMAP) {
    ENVIRONMENTMAP["ENVIRONMENTMAP_LOADED"] = "environmentmap.loaded";
})(ENVIRONMENTMAP = exports.ENVIRONMENTMAP || (exports.ENVIRONMENTMAP = {}));
var RENDERING;
(function (RENDERING) {
    RENDERING["BEAUTY_RENDERING_FINISHED"] = "rendering.beautyrenderingfinished";
})(RENDERING = exports.RENDERING || (exports.RENDERING = {}));
var SCENE;
(function (SCENE) {
    SCENE["SCENE_BOUNDING_BOX_CHANGE"] = "scene.boundingboxchange";
})(SCENE = exports.SCENE || (exports.SCENE = {}));
var VIEWER;
(function (VIEWER) {
    VIEWER["VIEWER_CREATED"] = "viewer.created";
    VIEWER["VIEWER_INITIALIZED"] = "viewer.initialized";
    VIEWER["VIEWER_UPDATED"] = "viewer.updated";
    VIEWER["VIEWER_CLOSED"] = "viewer.closed";
})(VIEWER = exports.VIEWER || (exports.VIEWER = {}));
var SESSION;
(function (SESSION) {
    SESSION["SESSION_CREATED"] = "session.created";
    SESSION["SESSION_INITIALIZED"] = "session.initialized";
    SESSION["SESSION_LOADED"] = "session.loaded";
    SESSION["SESSION_CUSTOMIZED"] = "session.customized";
    SESSION["SESSION_CLOSED"] = "session.closed";
})(SESSION = exports.SESSION || (exports.SESSION = {}));
var SETTINGS;
(function (SETTINGS) {
    SETTINGS["SETTINGS_REGISTERED"] = "settings.registered";
    SETTINGS["SETTINGS_REGISTERED_EXTERNAL"] = "settings.registered.external";
    SETTINGS["SETTINGS_UPDATE"] = "settings.update";
})(SETTINGS = exports.SETTINGS || (exports.SETTINGS = {}));
exports.EVENTTYPE = {
    CAMERA,
    ENVIRONMENTMAP,
    RENDERING,
    SCENE,
    SESSION,
    SETTINGS,
    VIEWER
};
