// @shapediver/viewer/dist/implementation/viewer/camera/AbstractCamera.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AbstractCamera_camera, _AbstractCamera_inputValidator, _AbstractCamera_logger, _AbstractCamera_tree, _AbstractCamera_viewer;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractCamera = void 0;
const tsyringe_1 = require("tsyringe");
const gl_matrix_1 = require("gl-matrix");
const viewer_shared_services_1 = require("@shapediver/viewer.shared.services");
const viewer_shared_math_1 = require("@shapediver/viewer.shared.math");
const viewer_shared_node_tree_1 = require("@shapediver/viewer.shared.node-tree");
class AbstractCamera {
    constructor(camera, viewer) {
        _AbstractCamera_camera.set(this, void 0);
        _AbstractCamera_inputValidator.set(this, tsyringe_1.container.resolve(viewer_shared_services_1.InputValidator));
        _AbstractCamera_logger.set(this, tsyringe_1.container.resolve(viewer_shared_services_1.Logger));
        _AbstractCamera_tree.set(this, tsyringe_1.container.resolve(viewer_shared_node_tree_1.Tree));
        _AbstractCamera_viewer.set(this, void 0);
        try {
            __classPrivateFieldSet(this, _AbstractCamera_camera, camera, "f");
            __classPrivateFieldSet(this, _AbstractCamera_viewer, viewer, "f");
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).constructor: Camera api created.`);
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).constructor: Something unexpected happened.`, true);
        }
    }
    set autoAdjust(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).autoAdjust: Updating AutoAdjust to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).autoAdjust`, value, 'boolean');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").autoAdjust = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).autoAdjust: autoAdjust was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).autoAdjust: Something unexpected happened.`, true);
        }
    }
    get autoAdjust() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").autoAdjust;
    }
    get cameraMovementDuration() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").cameraMovementDuration;
    }
    set cameraMovementDuration(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateCameraMovementDuration: Updating CameraMovementDuration to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateCameraMovementDuration`, value, 'positive');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").cameraMovementDuration = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateCameraMovementDuration: cameraMovementDuration was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updateCameraMovementDuration: Something unexpected happened.`, true);
        }
    }
    get defaultPosition() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").defaultPosition;
    }
    set defaultPosition(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateDefaultPosition: Updating DefaultPosition to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateDefaultPosition`, value, 'vec3');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").defaultPosition = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateDefaultPosition: defaultPosition was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updateDefaultPosition: Something unexpected happened.`, true);
        }
    }
    get defaultTarget() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").defaultTarget;
    }
    set defaultTarget(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateDefaultTarget: Updating DefaultTarget to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateDefaultTarget`, value, 'vec3');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").defaultTarget = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateDefaultTarget: defaultTarget was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updateDefaultTarget: Something unexpected happened.`, true);
        }
    }
    get enableCameraControls() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").enableCameraControls;
    }
    set enableCameraControls(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateEnableCameraControls: Updating EnableCameraControls to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateEnableCameraControls`, value, 'boolean');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").enableCameraControls = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateEnableCameraControls: enableCameraControls was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updateEnableCameraControls: Something unexpected happened.`, true);
        }
    }
    get id() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").id;
    }
    get order() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").order;
    }
    set order(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateOrder: Updating Order to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateOrder`, value, 'number', false);
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").order = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateOrder: order was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updateOrder: Something unexpected happened.`, true);
        }
    }
    get position() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").position;
    }
    set position(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updatePosition: Updating Position to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updatePosition`, value, 'vec3');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").position = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updatePosition: position was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updatePosition: Something unexpected happened.`, true);
        }
    }
    get revertAtMouseUp() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").revertAtMouseUp;
    }
    set revertAtMouseUp(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateRevertAtMouseUp: Updating RevertAtMouseUp to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateRevertAtMouseUp`, value, 'boolean');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").revertAtMouseUp = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateRevertAtMouseUp: revertAtMouseUp was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updateRevertAtMouseUp: Something unexpected happened.`, true);
        }
    }
    get revertAtMouseUpDuration() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").revertAtMouseUpDuration;
    }
    set revertAtMouseUpDuration(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateRevertAtMouseUpDuration: Updating RevertAtMouseUpDuration to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateRevertAtMouseUpDuration`, value, 'positive');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").revertAtMouseUpDuration = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateRevertAtMouseUpDuration: revertAtMouseUpDuration was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updateRevertAtMouseUpDuration: Something unexpected happened.`, true);
        }
    }
    get target() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").target;
    }
    set target(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateTarget: Updating Target to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateTarget`, value, 'vec3');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").target = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateTarget: target was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updateTarget: Something unexpected happened.`, true);
        }
    }
    get type() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").type;
    }
    get zoomExtentsFactor() {
        return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").zoomExtentsFactor;
    }
    set zoomExtentsFactor(value) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateZoomExtentsFactor: Updating ZoomExtentsFactor to ${value}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateZoomExtentsFactor`, value, 'positive');
            __classPrivateFieldGet(this, _AbstractCamera_camera, "f").zoomExtentsFactor = value;
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).updateZoomExtentsFactor: zoomExtentsFactor was set to: ${value}`);
            __classPrivateFieldGet(this, _AbstractCamera_viewer, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).updateZoomExtentsFactor: Something unexpected happened.`, true);
        }
    }
    animate(path, options) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate: Animating with path ${path} and options ${JSON.stringify(options)}.`);
            for (let i = 0; i < path.length; i++) {
                __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate`, path[i].position, 'vec3');
                __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate`, path[i].target, 'vec3');
            }
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate`, options, 'object', false);
            const o = Object.assign({}, options);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate`, o.easing, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate`, o.duration, 'number', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate`, o.default, 'boolean', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate`, o.coordinates, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate`, o.interpolation, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).animate: Starting camera path animation.`);
            return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").animate(path, o);
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).animate: Something unexpected happened.`, true);
        }
    }
    reset(options) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).reset: Resetting with options ${JSON.stringify(options)}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).reset`, options, 'object', false);
            const o = Object.assign({}, options);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).reset`, o.easing, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).reset`, o.duration, 'number', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).reset`, o.coordinates, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).reset`, o.interpolation, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).reset: Resetting position and target.`);
            return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").reset(o);
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).reset: Something unexpected happened.`, true);
        }
    }
    set(position, target, options) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).set: Setting to position ${position} and target ${target} with options ${JSON.stringify(options)}.`);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).set`, position, 'vec3');
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).set`, target, 'vec3');
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).set`, options, 'object', false);
            const o = Object.assign({}, options);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).set`, o.easing, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).set`, o.duration, 'number', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).set`, o.coordinates, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).set`, o.interpolation, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).set: Setting position to ${position} and target to ${target}.`);
            return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").set(gl_matrix_1.vec3.fromValues(position[0], position[1], position[2]), gl_matrix_1.vec3.fromValues(target[0], target[1], target[2]), o);
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).set: Something unexpected happened.`, true);
        }
    }
    zoomTo(zoomTarget, options) {
        try {
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").debugLow(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).zoomTo: Zooming to ${zoomTarget} with options ${JSON.stringify(options)}.`);
            let target;
            if (zoomTarget) {
                if (Array.isArray(zoomTarget)) {
                    __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).zoomTo`, zoomTarget, 'stringArray');
                    target = new viewer_shared_math_1.Box();
                    for (let i = 0; i < zoomTarget.length; i++) {
                        const node = __classPrivateFieldGet(this, _AbstractCamera_tree, "f").getNodeAtPath(zoomTarget[i]);
                        if (node)
                            target.union(node.boundingBox);
                        console.log(target);
                    }
                }
                else if (zoomTarget instanceof viewer_shared_math_1.Box) {
                    target = zoomTarget.clone();
                }
                else {
                    __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, new viewer_shared_services_1.SDError(`Camera(${this.id}).zoomTo: The specified zoom target does not have a valid type`));
                }
            }
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).zoomTo`, options, 'object', false);
            const o = Object.assign({}, options);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).zoomTo`, o.easing, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).zoomTo`, o.duration, 'number', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).zoomTo`, o.coordinates, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).zoomTo`, o.interpolation, 'string', false);
            __classPrivateFieldGet(this, _AbstractCamera_logger, "f").info(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, `Camera(${this.id}).zoomTo: Zooming in.`);
            return __classPrivateFieldGet(this, _AbstractCamera_camera, "f").zoomTo(target, o);
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.SDError)
                throw e;
            throw __classPrivateFieldGet(this, _AbstractCamera_logger, "f").error(viewer_shared_services_1.LOGGINGTOPIC.CAMERA, e, `Camera(${this.id}).zoomTo: Something unexpected happened.`, true);
        }
    }
}
exports.AbstractCamera = AbstractCamera;
_AbstractCamera_camera = new WeakMap(), _AbstractCamera_inputValidator = new WeakMap(), _AbstractCamera_logger = new WeakMap(), _AbstractCamera_tree = new WeakMap(), _AbstractCamera_viewer = new WeakMap();
