// @shapediver/api.geometry-api-dto-v1/dist/response/SdResponseDtoModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverResponseCollectionItemLinkRel = exports.ShapeDiverResponseCollectionItemDataName = void 0;
var ShapeDiverResponseCollectionItemDataName;
(function (ShapeDiverResponseCollectionItemDataName) {
    ShapeDiverResponseCollectionItemDataName["ACCESSDOMAINS"] = "accessdomains";
    ShapeDiverResponseCollectionItemDataName["AUTH_GROUPS"] = "auth_groups";
    ShapeDiverResponseCollectionItemDataName["BACKENDACCESS"] = "backendaccess";
    ShapeDiverResponseCollectionItemDataName["BACKLINKURL"] = "backlinkurl";
    ShapeDiverResponseCollectionItemDataName["CHECKURL"] = "checkurl";
    ShapeDiverResponseCollectionItemDataName["COMP_TIME"] = "comptime";
    ShapeDiverResponseCollectionItemDataName["CREATEDATE"] = "createdate";
    ShapeDiverResponseCollectionItemDataName["DENY_SCRIPTS"] = "deny_scripts";
    ShapeDiverResponseCollectionItemDataName["DOWNLOADURL"] = "downloadurl";
    ShapeDiverResponseCollectionItemDataName["FILENAME"] = "filename";
    ShapeDiverResponseCollectionItemDataName["FTYPE"] = "ftype";
    ShapeDiverResponseCollectionItemDataName["ID"] = "id";
    ShapeDiverResponseCollectionItemDataName["ID2"] = "id2";
    ShapeDiverResponseCollectionItemDataName["INITIAL_WARMUP"] = "initial_warmup";
    ShapeDiverResponseCollectionItemDataName["LAST_SESSION"] = "lastsession";
    ShapeDiverResponseCollectionItemDataName["LAST_VIEW"] = "lastview";
    ShapeDiverResponseCollectionItemDataName["MAX_COMP_TIME"] = "max_comp_time";
    ShapeDiverResponseCollectionItemDataName["MAX_EXPORT_SIZE"] = "max_export_size";
    ShapeDiverResponseCollectionItemDataName["MAX_IDLE_MINUTES"] = "max_idle_minutes";
    ShapeDiverResponseCollectionItemDataName["MAX_MODEL_SIZE"] = "max_model_size";
    ShapeDiverResponseCollectionItemDataName["MAX_OUTPUT_SIZE"] = "max_output_size";
    ShapeDiverResponseCollectionItemDataName["MAX_WAIT_TIME"] = "max_wait_time";
    ShapeDiverResponseCollectionItemDataName["MEM_USAGE"] = "memUsage";
    ShapeDiverResponseCollectionItemDataName["MSG"] = "msg";
    ShapeDiverResponseCollectionItemDataName["NAME"] = "name";
    ShapeDiverResponseCollectionItemDataName["NUM_COMP"] = "numcomp";
    ShapeDiverResponseCollectionItemDataName["NUM_LOADED_MAX"] = "num_loaded_max";
    ShapeDiverResponseCollectionItemDataName["NUM_LOADED_MIN"] = "num_loaded_min";
    ShapeDiverResponseCollectionItemDataName["NUM_PRELOADED_MIN"] = "num_preloaded_min";
    ShapeDiverResponseCollectionItemDataName["NUM_SESSIONS"] = "numsessions";
    ShapeDiverResponseCollectionItemDataName["PUB"] = "pub";
    ShapeDiverResponseCollectionItemDataName["REQUEST_TIME"] = "requesttime";
    ShapeDiverResponseCollectionItemDataName["REQUIRE_TOKEN"] = "require_token";
    ShapeDiverResponseCollectionItemDataName["S3URL"] = "s3url";
    ShapeDiverResponseCollectionItemDataName["SESSION_RATE_LIMIT"] = "session_rate_limit";
    ShapeDiverResponseCollectionItemDataName["SIZE"] = "size";
    ShapeDiverResponseCollectionItemDataName["STAT"] = "stat";
    ShapeDiverResponseCollectionItemDataName["TICKET"] = "ticket";
    ShapeDiverResponseCollectionItemDataName["TRUST"] = "trust";
    ShapeDiverResponseCollectionItemDataName["WEBHOOK_URL"] = "webhook_url";
    ShapeDiverResponseCollectionItemDataName["WEBHOOK_TOKEN"] = "webhook_token";
})(ShapeDiverResponseCollectionItemDataName = exports.ShapeDiverResponseCollectionItemDataName || (exports.ShapeDiverResponseCollectionItemDataName = {}));
var ShapeDiverResponseCollectionItemLinkRel;
(function (ShapeDiverResponseCollectionItemLinkRel) {
    ShapeDiverResponseCollectionItemLinkRel["TICKET"] = "ticket";
})(ShapeDiverResponseCollectionItemLinkRel = exports.ShapeDiverResponseCollectionItemLinkRel || (exports.ShapeDiverResponseCollectionItemLinkRel = {}));
