// @shapediver/viewer.settings/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.evaluateSettingsVersion = exports.validate = exports.convert = exports.DefaultsV3 = exports.DefaultsV2 = exports.DefaultsV1 = void 0;
const Defaults_1 = require("./versions/v1/Defaults");
Object.defineProperty(exports, "DefaultsV1", { enumerable: true, get: function () { return Defaults_1.Defaults; } });
const Validator_1 = require("./versions/v1/Validator");
const Defaults_2 = require("./versions/v2/Defaults");
Object.defineProperty(exports, "DefaultsV2", { enumerable: true, get: function () { return Defaults_2.Defaults; } });
const Validator_2 = require("./versions/v2/Validator");
const Converter_1 = require("./versions/v2/Converter");
const Defaults_3 = require("./versions/v3/Defaults");
Object.defineProperty(exports, "DefaultsV3", { enumerable: true, get: function () { return Defaults_3.Defaults; } });
const Validator_3 = require("./versions/v3/Validator");
const Converter_2 = require("./versions/v3/Converter");
let settingsUtilities = [];
settingsUtilities.push({
    version: '1.0',
    defaults: Defaults_1.Defaults,
    convertToPrevious: s => s,
    convertFromPrevious: s => s,
    validate: Validator_1.validate
});
settingsUtilities.push({
    version: '2.0',
    defaults: Defaults_2.Defaults,
    convertToPrevious: Converter_1.convertToPrevious,
    convertFromPrevious: Converter_1.convertFromPrevious,
    validate: Validator_2.validate
});
settingsUtilities.push({
    version: '3.0',
    defaults: Defaults_3.Defaults,
    convertToPrevious: Converter_2.convertToPrevious,
    convertFromPrevious: Converter_2.convertFromPrevious,
    validate: Validator_3.validate
});
const convert = (settings, targetVersion) => {
    if (settings.settings_version === targetVersion)
        return settings;
    const target = settingsUtilities.findIndex(util => { return util.version === targetVersion; });
    const current = settingsUtilities.findIndex(util => { return util.version === settings.settings_version; });
    if (target === -1)
        throw new Error('ViewerSettings.convert: Target version not available');
    if (current === -1)
        throw new Error('ViewerSettings.convert: Settings version not available');
    let tempSettings = settings;
    if (target < current) {
        for (let i = current; target < i; i--)
            tempSettings = settingsUtilities[i].convertToPrevious(tempSettings);
    }
    else {
        for (let i = current + 1; i <= target; i++)
            tempSettings = settingsUtilities[i].convertFromPrevious(tempSettings);
    }
    return tempSettings;
};
exports.convert = convert;
const validate = (settings, targetVersion) => {
    if (targetVersion !== undefined) {
        const index = settingsUtilities.findIndex(util => { return util.version === targetVersion; });
        if (index === -1)
            throw new Error('ViewerSettings.validate: Target version was not found.');
        if (settings.settings_version !== undefined && settings.settings_version !== targetVersion)
            throw new Error('ViewerSettings.validate: The settings do have a different version than the target version.');
        settingsUtilities[index].validate(settings);
    }
    else {
        if (!settings.settings_version)
            throw new Error('ViewerSettings.validate: Settings do not have a version specified.');
        const index = settingsUtilities.findIndex(util => { return util.version === settings.settings_version; });
        settingsUtilities[index].validate(settings);
    }
};
exports.validate = validate;
const evaluateSettingsVersion = (viewerVersion) => {
    if (viewerVersion.startsWith('0') || viewerVersion === '1.0.0' || viewerVersion === '1.0.1' || viewerVersion === '1.0.2' || viewerVersion === '1.0.3' || viewerVersion === '1.0.4' || viewerVersion === '1.0.5')
        return '2.0';
    if (viewerVersion.startsWith('1'))
        return '1.0';
    if (viewerVersion.startsWith('2'))
        return '2.0';
    return '3.0';
};
exports.evaluateSettingsVersion = evaluateSettingsVersion;
